### PR TITLE
Data format specifications

### PR DESCRIPTION
**Pull request by sstucker**
The current version of the spec leaves far too much ambiguity in its descriptions of the way HDF5 Datasets should represent SNIRF objects.

This has led to hiccups in SNIRF interface implementations:

* SNIRF files written by Kernel engineers had string values stored in a then-unfamiliar "variable length" format which has consequences for the HDF5 interface as it is implemented in MATLAB, breaking read operations in Homer3. (#72) The language of the specification does not distinguish between this format nor the fixed-length format: as it stands, both are valid.
* SNIRF files written via a naive usage of the MATLAB HDF5 interface do not use the [scalar dataspace](http://davis.lbl.gov/Manuals/HDF5-1.8.7/UG/UG_frame12Dataspaces.html), instead saving every Dataset as a simple dataspace (interchangeable with "array" in terms of HDF5 formatting) of `rank`=1 and `size`=[1]. In contrast, Python's `h5py` interface does save single values and strings using the `scalar dataspace`. (#98) The current langauge of the spec does not make any mention of HDF5 dataspaces and therefore does not distinguish between these formats. Consequently, read operations on SNIRF files must expect any `numeric`, `string` or `integer` value to appear with a size and dimension equivalent to that of a `1-D array` of the same type, or in very different `scalar` format. This makes read-only validation of SNIRF files difficult and complicates read logic.

This PR aims to further specify the formats of SNIRF fields in terms of HDF5 dataspaces in an elaborated "Data format" section. I have also added code samples in both MATLAB and Python which demonstrate the specified way to write datasets to ease programmatic implementation of this breaking change.

Use of the "low-level" interface is unfortunately necessary to properly write strings and scalar Datasets in a MATLAB environment.

I have created and added the breaking label to signify that this constitutes a breaking change to previous SNIRF implementations.

### CODE REVIEW

**Comment by sstucker** [comment]
As this is the first breaking change proposed, we could discuss a deprecation process.

I feel okay about a breaking release, as long we provide code to users to meet the specification or convert to valid format (I have, in this PR and with pysnirf2) and can assume that most implementations will be backwards compatible.

**Comment by sstucker** [comment]
It would also be great to have someone independently test and verify the code samples in both MATLAB and Python and confirm they work... want to ping @andyzjc

**Comment by chrsthur** [comment]
> It would also be great to have someone independently test and verify the code samples in both MATLAB and Python and confirm they work... pinging @andyzjc and @chrsthur

Hi Stephen,
Which code samples specifically are you referring to here?

**Comment by sstucker** [comment]
> Hi Stephen, Which code samples specifically are you referring to here?

https://github.com/sstucker/snirf/blob/formatting/snirf_specification.md#code-samples

These demonstrate saving of SNIRF fields such that they meet the new specifications

**Comment by sstucker** [comment]
It might make sense to move the code samples out of the spec document and into a separate document

**Comment by dboas** [comment]
We have to implement this fix. It will ensure all future SNIRF files are the same. And it seems @sstucker has already made the code examples to fix broken snirf files.

File: snirf_specification.md [hunk 1]
```diff
@@ -969,6 +971,118 @@ would be specified as follows:
   [0.6 0.2 1.0]
 ```
 
+### Code samples
```
Comment (rob-luke) [inline snirf_specification.md:974]: This is awesome!

**Review by rob-luke — APPROVED** [review]
Great stuff @sstucker

Is this a breaking change for the spec? Or is it just clarifying what was already specified? Previously the spec said that a field must be an integer, and that is still the case. You haven't changed any of the field types. Never did it previously say that an integer was actually a single-element array, nor would it be intuitive to have thought that (because there was an array type if that had been required). This PR makes it explicit that an integer must be an integer, etc. If some software had stored integers as arrays, I'd say that's an implementation error and not a fault of the spec. I raise this because if you believe this is a breaking change we would increment the version to `v2.0.0`, no big issue if needed, but is it needed?

**Comment by sstucker** [comment]
> Is this a breaking change for the spec?

I like your optimistic definition of breaking and I suppose I agree-- I consider it breaking because people's files (Homer's, in fact) are invalid version 1.x SNIRF files once we release this. But maybe that is not a good definition.

**Comment by sstucker** [comment]
I have verified these with the help of @chrsthur

**Pull request merged**